In [1]:
import pandas as pd
from pandasai import SmartDataframe

In [2]:
from pandasai.llm import OpenAI
llm = OpenAI(api_token = "Insertar API Token")

In [3]:
df = pd.read_csv("student.csv", sep = ",")
df.head()

Unnamed: 0    Id  Student_Age     Sex High_School_Type Scholarship  \
0           0  5001           21    Male            Other         50%   
1           1  5002           20    Male            Other         50%   
2           2  5003           21    Male            State         50%   
3           3  5004           18  Female          Private         50%   
4           4  5005           22    Male          Private         50%   

  Additional_Work Sports_activity Transportation  Weekly_Study_Hours  \
0             Yes              No        Private                   0   
1             Yes              No        Private                   0   
2              No              No        Private                   2   
3             Yes              No            Bus                   2   
4              No              No            Bus                  12   

  Attendance Reading Notes Listening_in_Class Project_work Grade  
0     Always     Yes   Yes                 No           No    AA  
1     Always     Yes    No                Yes          Yes    AA  
2      Never      No    No                 No          Yes    AA  
3     Always      No   Yes                 No           No    AA  
4     Always     Yes    No                Yes          Yes    AA

Iniciar df

In [4]:
sdf = SmartDataframe(df, config={"llm": llm, "verbose": True})

EDA

In [5]:
print("Dataframe shape: ", df.shape)
print("Dataframe columns: ", df.columns)
print("Dataframe info: ", df.info())
print("Dataframe describe: ", df.describe())
print("Dataframe null values: ", df.isnull().sum())

Dataframe shape:  (145, 16)
Dataframe columns:  Index(['Unnamed: 0', 'Id', 'Student_Age', 'Sex', 'High_School_Type',
       'Scholarship', 'Additional_Work', 'Sports_activity', 'Transportation',
       'Weekly_Study_Hours', 'Attendance', 'Reading', 'Notes',
       'Listening_in_Class', 'Project_work', 'Grade'],
      dtype='object')
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 145 entries, 0 to 144
Data columns (total 16 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   Unnamed: 0          145 non-null    int64 
 1   Id                  145 non-null    int64 
 2   Student_Age         145 non-null    int64 
 3   Sex                 145 non-null    object
 4   High_School_Type    145 non-null    object
 5   Scholarship         145 non-null    object
 6   Additional_Work     145 non-null    object
 7   Sports_activity     145 non-null    object
 8   Transportation      145 non-null    object
 9   Weekly_Study_Hours  145 non

In [6]:
sdf.chat("What is the average age of students?")

2024-04-17 22:49:32 [INFO] Question: What is the average age of students?
2024-04-17 22:49:33 [INFO] Running PandasAI with openai LLM...
2024-04-17 22:49:33 [INFO] Prompt ID: 4075d370-198e-4b31-804d-b8c992a216dc
2024-04-17 22:49:33 [INFO] Executing Pipeline: GenerateChatPipeline
2024-04-17 22:49:33 [INFO] Executing Step 0: ValidatePipelineInput
2024-04-17 22:49:33 [INFO] Executing Step 1: CacheLookup
2024-04-17 22:49:33 [INFO] Executing Step 2: PromptGeneration
2024-04-17 22:49:34 [INFO] Using prompt: dfs[0]:
  name: null
  description: null
  type: pd.DataFrame
  rows: 145
  columns: 16
  schema:
    fields:
    - name: 'Unnamed: 0'
      type: int64
      samples:
      - 134
      - 67
      - 103
    - name: Id
      type: int64
      samples:
      - 5064
      - 5145
      - 5079
    - name: Student_Age
      type: int64
      samples:
      - 25
      - 22
      - 24
    - name: Sex
      type: object
      samples:
      - Male
      - Male
      - Female
    - name: High_Schoo

Traceback (most recent call last):
  File "c:\Users\maria\AppData\Local\Programs\Python\Python311\Lib\site-packages\pandasai\pipelines\chat\generate_chat_pipeline.py", line 281, in run
    output = (self.code_generation_pipeline | self.code_execution_pipeline).run(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "c:\Users\maria\AppData\Local\Programs\Python\Python311\Lib\site-packages\pandasai\pipelines\pipeline.py", line 137, in run
    raise e
  File "c:\Users\maria\AppData\Local\Programs\Python\Python311\Lib\site-packages\pandasai\pipelines\pipeline.py", line 101, in run
    step_output = logic.execute(
                  ^^^^^^^^^^^^^^
  File "c:\Users\maria\AppData\Local\Programs\Python\Python311\Lib\site-packages\pandasai\pipelines\chat\code_generator.py", line 33, in execute
    code = pipeline_context.config.llm.generate_code(input, pipeline_context)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "c:\U

"Unfortunately, I was not able to answer your question, because of the following error:\n\nError code: 401 - {'error': {'message': 'Incorrect API key provided: Insertar******oken. You can find your API key at https://platform.openai.com/account/api-keys.', 'type': 'invalid_request_error', 'param': None, 'code': 'invalid_api_key'}}\n"

In [7]:
sdf.chat("Graficar la distribución de las edades de los estudiantes")

2024-04-17 22:49:34 [INFO] Question: Graficar la distribución de las edades de los estudiantes
2024-04-17 22:49:34 [INFO] Running PandasAI with openai LLM...
2024-04-17 22:49:34 [INFO] Prompt ID: 2084c36b-d50d-4c30-9d0c-f714bc41d351
2024-04-17 22:49:34 [INFO] Executing Pipeline: GenerateChatPipeline
2024-04-17 22:49:34 [INFO] Executing Step 0: ValidatePipelineInput
2024-04-17 22:49:34 [INFO] Executing Step 1: CacheLookup
2024-04-17 22:49:35 [INFO] Executing Step 2: PromptGeneration
2024-04-17 22:49:35 [INFO] Using prompt: dfs[0]:
  name: null
  description: null
  type: pd.DataFrame
  rows: 145
  columns: 16
  schema:
    fields:
    - name: 'Unnamed: 0'
      type: int64
      samples:
      - 134
      - 67
      - 103
    - name: Id
      type: int64
      samples:
      - 5064
      - 5145
      - 5079
    - name: Student_Age
      type: int64
      samples:
      - 25
      - 22
      - 24
    - name: Sex
      type: object
      samples:
      - Male
      - Male
      - Female
 

Traceback (most recent call last):
  File "c:\Users\maria\AppData\Local\Programs\Python\Python311\Lib\site-packages\pandasai\pipelines\chat\generate_chat_pipeline.py", line 281, in run
    output = (self.code_generation_pipeline | self.code_execution_pipeline).run(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "c:\Users\maria\AppData\Local\Programs\Python\Python311\Lib\site-packages\pandasai\pipelines\pipeline.py", line 137, in run
    raise e
  File "c:\Users\maria\AppData\Local\Programs\Python\Python311\Lib\site-packages\pandasai\pipelines\pipeline.py", line 101, in run
    step_output = logic.execute(
                  ^^^^^^^^^^^^^^
  File "c:\Users\maria\AppData\Local\Programs\Python\Python311\Lib\site-packages\pandasai\pipelines\chat\code_generator.py", line 33, in execute
    code = pipeline_context.config.llm.generate_code(input, pipeline_context)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "c:\U

"Unfortunately, I was not able to answer your question, because of the following error:\n\nError code: 401 - {'error': {'message': 'Incorrect API key provided: Insertar******oken. You can find your API key at https://platform.openai.com/account/api-keys.', 'type': 'invalid_request_error', 'param': None, 'code': 'invalid_api_key'}}\n"

In [8]:
sdf.chat("¿Cuál es la edad promedio de los estudiantes que son mujeres?")

2024-04-17 22:49:35 [INFO] Question: ¿Cuál es la edad promedio de los estudiantes que son mujeres?
2024-04-17 22:49:35 [INFO] Running PandasAI with openai LLM...
2024-04-17 22:49:35 [INFO] Prompt ID: 2b1b899a-5771-4484-908c-da11f7c27d84
2024-04-17 22:49:35 [INFO] Executing Pipeline: GenerateChatPipeline
2024-04-17 22:49:35 [INFO] Executing Step 0: ValidatePipelineInput
2024-04-17 22:49:35 [INFO] Executing Step 1: CacheLookup
2024-04-17 22:49:35 [INFO] Executing Step 2: PromptGeneration
2024-04-17 22:49:35 [INFO] Using prompt: dfs[0]:
  name: null
  description: null
  type: pd.DataFrame
  rows: 145
  columns: 16
  schema:
    fields:
    - name: 'Unnamed: 0'
      type: int64
      samples:
      - 134
      - 67
      - 103
    - name: Id
      type: int64
      samples:
      - 5064
      - 5145
      - 5079
    - name: Student_Age
      type: int64
      samples:
      - 25
      - 22
      - 24
    - name: Sex
      type: object
      samples:
      - Male
      - Male
      - Fema

Traceback (most recent call last):
  File "c:\Users\maria\AppData\Local\Programs\Python\Python311\Lib\site-packages\pandasai\pipelines\chat\generate_chat_pipeline.py", line 281, in run
    output = (self.code_generation_pipeline | self.code_execution_pipeline).run(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "c:\Users\maria\AppData\Local\Programs\Python\Python311\Lib\site-packages\pandasai\pipelines\pipeline.py", line 137, in run
    raise e
  File "c:\Users\maria\AppData\Local\Programs\Python\Python311\Lib\site-packages\pandasai\pipelines\pipeline.py", line 101, in run
    step_output = logic.execute(
                  ^^^^^^^^^^^^^^
  File "c:\Users\maria\AppData\Local\Programs\Python\Python311\Lib\site-packages\pandasai\pipelines\chat\code_generator.py", line 33, in execute
    code = pipeline_context.config.llm.generate_code(input, pipeline_context)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "c:\U

"Unfortunately, I was not able to answer your question, because of the following error:\n\nError code: 401 - {'error': {'message': 'Incorrect API key provided: Insertar******oken. You can find your API key at https://platform.openai.com/account/api-keys.', 'type': 'invalid_request_error', 'param': None, 'code': 'invalid_api_key'}}\n"

In [9]:
sdf.chat("¿Cuál es la edad promedio de los estudiantes que son hombres?")

2024-04-17 22:49:36 [INFO] Question: ¿Cuál es la edad promedio de los estudiantes que son hombres?
2024-04-17 22:49:36 [INFO] Running PandasAI with openai LLM...
2024-04-17 22:49:36 [INFO] Prompt ID: 3c68ce90-73b5-4d8d-8ea2-659fbfe1677a
2024-04-17 22:49:36 [INFO] Executing Pipeline: GenerateChatPipeline
2024-04-17 22:49:36 [INFO] Executing Step 0: ValidatePipelineInput
2024-04-17 22:49:36 [INFO] Executing Step 1: CacheLookup
2024-04-17 22:49:36 [INFO] Executing Step 2: PromptGeneration
2024-04-17 22:49:36 [INFO] Using prompt: dfs[0]:
  name: null
  description: null
  type: pd.DataFrame
  rows: 145
  columns: 16
  schema:
    fields:
    - name: 'Unnamed: 0'
      type: int64
      samples:
      - 134
      - 67
      - 103
    - name: Id
      type: int64
      samples:
      - 5064
      - 5145
      - 5079
    - name: Student_Age
      type: int64
      samples:
      - 25
      - 22
      - 24
    - name: Sex
      type: object
      samples:
      - Male
      - Male
      - Fema

Traceback (most recent call last):
  File "c:\Users\maria\AppData\Local\Programs\Python\Python311\Lib\site-packages\pandasai\pipelines\chat\generate_chat_pipeline.py", line 281, in run
    output = (self.code_generation_pipeline | self.code_execution_pipeline).run(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "c:\Users\maria\AppData\Local\Programs\Python\Python311\Lib\site-packages\pandasai\pipelines\pipeline.py", line 137, in run
    raise e
  File "c:\Users\maria\AppData\Local\Programs\Python\Python311\Lib\site-packages\pandasai\pipelines\pipeline.py", line 101, in run
    step_output = logic.execute(
                  ^^^^^^^^^^^^^^
  File "c:\Users\maria\AppData\Local\Programs\Python\Python311\Lib\site-packages\pandasai\pipelines\chat\code_generator.py", line 33, in execute
    code = pipeline_context.config.llm.generate_code(input, pipeline_context)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "c:\U

"Unfortunately, I was not able to answer your question, because of the following error:\n\nError code: 401 - {'error': {'message': 'Incorrect API key provided: Insertar******oken. You can find your API key at https://platform.openai.com/account/api-keys.', 'type': 'invalid_request_error', 'param': None, 'code': 'invalid_api_key'}}\n"

In [10]:
sdf.chat("¿Cuál es la edad promedio de los estudiantes que son hombres y que tienen una calificación mayor a 70?")

2024-04-17 22:49:36 [INFO] Question: ¿Cuál es la edad promedio de los estudiantes que son hombres y que tienen una calificación mayor a 70?
2024-04-17 22:49:36 [INFO] Running PandasAI with openai LLM...
2024-04-17 22:49:36 [INFO] Prompt ID: ff9ccc3f-fa43-4bb4-aa7d-a527dbe98a93
2024-04-17 22:49:37 [INFO] Executing Pipeline: GenerateChatPipeline
2024-04-17 22:49:37 [INFO] Executing Step 0: ValidatePipelineInput
2024-04-17 22:49:37 [INFO] Executing Step 1: CacheLookup
2024-04-17 22:49:37 [INFO] Executing Step 2: PromptGeneration
2024-04-17 22:49:37 [INFO] Using prompt: dfs[0]:
  name: null
  description: null
  type: pd.DataFrame
  rows: 145
  columns: 16
  schema:
    fields:
    - name: 'Unnamed: 0'
      type: int64
      samples:
      - 134
      - 67
      - 103
    - name: Id
      type: int64
      samples:
      - 5064
      - 5145
      - 5079
    - name: Student_Age
      type: int64
      samples:
      - 25
      - 22
      - 24
    - name: Sex
      type: object
      sample

Traceback (most recent call last):
  File "c:\Users\maria\AppData\Local\Programs\Python\Python311\Lib\site-packages\pandasai\pipelines\chat\generate_chat_pipeline.py", line 281, in run
    output = (self.code_generation_pipeline | self.code_execution_pipeline).run(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "c:\Users\maria\AppData\Local\Programs\Python\Python311\Lib\site-packages\pandasai\pipelines\pipeline.py", line 137, in run
    raise e
  File "c:\Users\maria\AppData\Local\Programs\Python\Python311\Lib\site-packages\pandasai\pipelines\pipeline.py", line 101, in run
    step_output = logic.execute(
                  ^^^^^^^^^^^^^^
  File "c:\Users\maria\AppData\Local\Programs\Python\Python311\Lib\site-packages\pandasai\pipelines\chat\code_generator.py", line 33, in execute
    code = pipeline_context.config.llm.generate_code(input, pipeline_context)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "c:\U

"Unfortunately, I was not able to answer your question, because of the following error:\n\nError code: 401 - {'error': {'message': 'Incorrect API key provided: Insertar******oken. You can find your API key at https://platform.openai.com/account/api-keys.', 'type': 'invalid_request_error', 'param': None, 'code': 'invalid_api_key'}}\n"

In [11]:
sdf.chat("De que tipo de high school son los estuaintes que tienen una calificación mayor a 70?")

2024-04-17 22:49:37 [INFO] Question: De que tipo de high school son los estuaintes que tienen una calificación mayor a 70?
2024-04-17 22:49:37 [INFO] Running PandasAI with openai LLM...
2024-04-17 22:49:37 [INFO] Prompt ID: a8cf1eb8-9846-4090-84c6-5a150601cce7
2024-04-17 22:49:37 [INFO] Executing Pipeline: GenerateChatPipeline
2024-04-17 22:49:38 [INFO] Executing Step 0: ValidatePipelineInput
2024-04-17 22:49:38 [INFO] Executing Step 1: CacheLookup
2024-04-17 22:49:38 [INFO] Executing Step 2: PromptGeneration
2024-04-17 22:49:38 [INFO] Using prompt: dfs[0]:
  name: null
  description: null
  type: pd.DataFrame
  rows: 145
  columns: 16
  schema:
    fields:
    - name: 'Unnamed: 0'
      type: int64
      samples:
      - 134
      - 67
      - 103
    - name: Id
      type: int64
      samples:
      - 5064
      - 5145
      - 5079
    - name: Student_Age
      type: int64
      samples:
      - 25
      - 22
      - 24
    - name: Sex
      type: object
      samples:
      - Male
 

Traceback (most recent call last):
  File "c:\Users\maria\AppData\Local\Programs\Python\Python311\Lib\site-packages\pandasai\pipelines\chat\generate_chat_pipeline.py", line 281, in run
    output = (self.code_generation_pipeline | self.code_execution_pipeline).run(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "c:\Users\maria\AppData\Local\Programs\Python\Python311\Lib\site-packages\pandasai\pipelines\pipeline.py", line 137, in run
    raise e
  File "c:\Users\maria\AppData\Local\Programs\Python\Python311\Lib\site-packages\pandasai\pipelines\pipeline.py", line 101, in run
    step_output = logic.execute(
                  ^^^^^^^^^^^^^^
  File "c:\Users\maria\AppData\Local\Programs\Python\Python311\Lib\site-packages\pandasai\pipelines\chat\code_generator.py", line 33, in execute
    code = pipeline_context.config.llm.generate_code(input, pipeline_context)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "c:\U

"Unfortunately, I was not able to answer your question, because of the following error:\n\nError code: 401 - {'error': {'message': 'Incorrect API key provided: Insertar******oken. You can find your API key at https://platform.openai.com/account/api-keys.', 'type': 'invalid_request_error', 'param': None, 'code': 'invalid_api_key'}}\n"

In [12]:
sdf.chat("Grafica de la validacion de datos atipicos del dataframe")

2024-04-17 22:49:38 [INFO] Question: Grafica de la validacion de datos atipicos del dataframe
2024-04-17 22:49:38 [INFO] Running PandasAI with openai LLM...
2024-04-17 22:49:38 [INFO] Prompt ID: e0900789-31f4-439a-9ab0-3a5623f9fbfa
2024-04-17 22:49:38 [INFO] Executing Pipeline: GenerateChatPipeline
2024-04-17 22:49:38 [INFO] Executing Step 0: ValidatePipelineInput
2024-04-17 22:49:38 [INFO] Executing Step 1: CacheLookup
2024-04-17 22:49:38 [INFO] Executing Step 2: PromptGeneration
2024-04-17 22:49:38 [INFO] Using prompt: dfs[0]:
  name: null
  description: null
  type: pd.DataFrame
  rows: 145
  columns: 16
  schema:
    fields:
    - name: 'Unnamed: 0'
      type: int64
      samples:
      - 134
      - 67
      - 103
    - name: Id
      type: int64
      samples:
      - 5064
      - 5145
      - 5079
    - name: Student_Age
      type: int64
      samples:
      - 25
      - 22
      - 24
    - name: Sex
      type: object
      samples:
      - Male
      - Male
      - Female
  

Traceback (most recent call last):
  File "c:\Users\maria\AppData\Local\Programs\Python\Python311\Lib\site-packages\pandasai\pipelines\chat\generate_chat_pipeline.py", line 281, in run
    output = (self.code_generation_pipeline | self.code_execution_pipeline).run(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "c:\Users\maria\AppData\Local\Programs\Python\Python311\Lib\site-packages\pandasai\pipelines\pipeline.py", line 137, in run
    raise e
  File "c:\Users\maria\AppData\Local\Programs\Python\Python311\Lib\site-packages\pandasai\pipelines\pipeline.py", line 101, in run
    step_output = logic.execute(
                  ^^^^^^^^^^^^^^
  File "c:\Users\maria\AppData\Local\Programs\Python\Python311\Lib\site-packages\pandasai\pipelines\chat\code_generator.py", line 33, in execute
    code = pipeline_context.config.llm.generate_code(input, pipeline_context)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "c:\U

"Unfortunately, I was not able to answer your question, because of the following error:\n\nError code: 401 - {'error': {'message': 'Incorrect API key provided: Insertar******oken. You can find your API key at https://platform.openai.com/account/api-keys.', 'type': 'invalid_request_error', 'param': None, 'code': 'invalid_api_key'}}\n"

### Informe

#### Exploración Inicial
Se realizó una exploración inicial del DataFrame para comprender su estructura y las características de los datos. Esto incluyó la verificación de las primeras filas del DataFrame, el número de filas y columnas, así como los tipos de datos de cada columna.

Resultados:
Número de Filas y Columnas: El DataFrame contiene 145 filas y 16 columnas.
Tipos de Datos: El DataFrame consta de 4 columnas numéricas y 12 columnas de tipo objeto (categóricas).
Resumen Estadístico: Se obtuvo un resumen estadístico que incluye recuentos, medias, desviaciones estándar y cuartiles para las columnas numéricas.


#### Análisis de Datos
Se realizaron análisis adicionales para comprender mejor los datos y extraer información relevante. Esto incluyó calcular el promedio de edad de los estudiantes, así como el promedio de edad para subgrupos específicos, como estudiantes femeninas y estudiantes masculinos con calificaciones superiores a 70.

Resultados:
Promedio de Edad de Estudiantes: El promedio de edad de todos los estudiantes es de 19.68 años.
Promedio de Edad de Estudiantes Females: El promedio de edad de las estudiantes femeninas es de 19.41 años.
Promedio de Edad de Estudiantes Masculinos con Calificaciones Superiores a 70: El promedio de edad de los estudiantes masculinos con calificaciones superiores a 70 es de 19.86 años.
Tipos de Escuela de Estudiantes con Calificaciones Superiores a 70: Los estudiantes que tienen una calificación superior a 70 asisten a los siguientes tipos de escuelas: Other, State, Private.

#### Conclusiones
El análisis exploratorio de datos proporcionó una comprensión inicial de las características del conjunto de datos de los estudiantes. Se identificaron algunos patrones interesantes relacionados con la edad de los estudiantes y su desempeño académico, así como el tipo de escuela al que asisten los estudiantes con calificaciones superiores a 70. 